software/py/vanilla_stats_parser.py

```diff
@@ -0,0 +1,952 @@
+                for miss in self.misses:
+                    tile_group_stat[tag][tg_id]["miss_total"] += tile_group_stat[tag][tg_id][miss]
+
+        self.instrs += ["instr_total"]
+        self.stalls += ["stall_total"]
+        self.misses += ["miss_total"]
+        self.all_ops += ["instr_total", "stall_total", "miss_total"]
+
+        return num_tags, num_tile_groups, tile_group_stat, tile_stat
+
+
+
+    # Generate a stats dictionary for each tile containing the stat and it's aggregate count
+    # other than timing, tile stats are only read once per tile from the end of file
+    # i.e. if mesh dimensions are 4x4, only last 16 lines are needed 
+    # Deprecated  -- might be used later if needed
+    # This method count the aggregate stats (including the time tiles are waiting
+    # for a program to be loaded)
+    def __generate_inclusive_tile_stat (self, traces):
+        tile_stat = [[Counter() for x in range(self.manycore_dim_x)] for y in range(self.manycore_dim_y)]
+        trace_idx = len(traces)
+        for y in range(self.manycore_dim_y):
+            for x in range(self.manycore_dim_x):
+                trace_idx -= 1
+                trace = traces[trace_idx]
+                for op in self.all_ops:
+                    tile_stat[y][x][op] = trace[op]
+        return tile_stat
+
+
+
+    # Calculate aggregate manycore stats dictionary by summing 
+    # all per tile stats dictionaries
+    def __generate_manycore_stats_all(self, tile_stat):
+        # Create a dictionary and initialize elements to zero
+        manycore_stat = [Counter() for tag in range(self.max_tags)]
+        for tag in range(self.max_tags):
+            for y in range(self.manycore_dim_y):
+                for x in range(self.manycore_dim_x):
+                    for op in self.all_ops:
+                        manycore_stat[tag][op] += tile_stat[tag][y][x][op]
+
+        return manycore_stat
+ 
+
+
+    # Parsers stat file's header to generate list of all 
+    # operations based on type (stat, instruction, miss, stall)
+    def parse_header(self, f):
+        # Generate lists of stats/instruction/miss/stall names
+        instrs = []
+        misses = []
+        stalls = []
+        stats  = []
+
+        with open(f) as fp:
+            rdr = csv.DictReader (fp, delimiter=",")
+      
+            header = rdr.fieldnames
+            for item in header:
+                if (item.startswith('instr_')):
+                    if (not item == 'instr_total'):
+                        instrs += [item]
+                elif (item.startswith('miss_')):
+                    misses += [item]
+                elif (item.startswith('stall_')):
+                    stalls += [item]
+                else:
+                    stats += [item]
+        return (stats, instrs, misses, stalls)
+
+
+
+    # main public method
+    # default stats generator
+    def generate_stats(self, input_file):
+
```

Comment:
I think that you could split this up, and remove it entirely. It's not a public method, so just parse the header (__parse_header) and then generate tile stats in init

Reply:
Done, I moved the contents inside the init function, how does it look now?